### DISCUSSION TITLE
Trouble Trying to Understand Config and 16bit CS line

### DISCUSSION THREAD

**Discussion by anatoli-dp** [lovyan03/LovyanGFX#607] [Device Configurations]
Hello all, I am working on a custom game console (BitByte Handheld Console if u wanna google it) and am working on a revision to interface an esp32s3 as a dedicated "gpu" to an ili9341 display. I want to try lovyangfx as I understand it supports 16bit parallel and ili9341 displays (have previously using TFT_eSPI before but it only supports 8bit parallel). I am having trouble understanding the configuration that is needed so that I can properly break out and route signals between the display and esp32s3.

The display being used -> https://www.alibaba.com/product-detail/2-8-inch-TFT-LCD-Screen_1600246407796.html (can also be found in adafruit stores as https://www.adafruit.com/product/2770)
The datasheet for the display being used -> https://cdn-shop.adafruit.com/product-files/2770/SPEC-CH280QV10-CT_Rev.D.pdf

My current pinout
![screen](https://github.com/user-attachments/assets/5860c581-7548-47b9-b66d-129c37943abf)

Now for the issue . . . Please note a lot of this is an educated guess as I am trying to understand the lovyan config so i can wire the board correctly in order to use it

In setup i would add
 >> lgfx::Panel_ILI9341 _panel_instance;
 >> lgfx::Bus_Parallel16 _bus_instance;
 >> pin_data[data0 -> data15]/rd/wr/rs

now the display in question also has a chip select which shows is used according to the datasheet
38 /CS Chip select input pin (“Low” enable) in MPU I/F and SPI I/F
does ili9341 use a chip select in 16bit parallel? If so what would it fall under for the config parameter? (asking this as previously only ever used spi and u had to have a chip select so unsure about 16 parallel)
and if not do I just ground the pin so it is always "active" or does 16bit not matter the state of cs and i can just leave it unconnected

is  "int8_t pin_ctrl[3] = { -1, -1, -1 };" something that is needed to be set manually or is this just a holder for something the library does? If it is manually set what are the control pins?

lovyan shows esp32s3 has 16bit hardware support though i thought it had to be done in software (as referenced by bodmer in a TFT_eSPI discussion surrounding 8bit clock speeds), so since it has hardware support is everything routed through the gpio mux or are there better defined pins for the output that are more "optimized" in regards to direct register mutation/io mux ecluded?

and lastly before i hardwire all this in does lovyan support pushing images from psram? This is just needed to determine if i set IO 35, 36, and 37 as the module has octal psram which i would love to try the higher clock frequency but if not supported then me using those pins for a more optimized layout is more valuable to me than having the higher speed that was introduced in idf 5 so very important to know before i make the board for routing

sorry for the questions and if they seem stupid, im just very slow at understanding things . . . and if u can help with the CS pin question especially regarding the display datasheet that would also be helpful as I am just unsure about that line if its not SPI (not that u have to since obviously that is more of a hardware question than library specific)

Thanks if you do decide to answer this

**Comment by tobozo** [comment]
hi,

> does ili9341 use a chip select in 16bit parallel? If so what would it fall under for the config parameter?

`cfg.pin_cs` goes in panel config, not bus config

```cpp
    {
      auto cfg = _panel_instance.config();
      cfg.pin_cs          =    TFT_CS;
      cfg.pin_rst         =    TFT_RST;
      cfg.pin_busy        =    -1;
      cfg.offset_rotation =     0;
      cfg.readable        =  true;
      cfg.invert          = false;
      cfg.rgb_order       = false;
      cfg.dlen_16bit      =  true;
      cfg.bus_shared      = false;
      _panel_instance.config(cfg);
    }
```

> is "int8_t pin_ctrl[3] = { -1, -1, -1 };" something that is needed to be set manually or is this just a holder for something the library does?

`cfg.pin_ctrl` is the [same](https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx/v1/platforms/esp32s3/Bus_Parallel16.hpp#L47-L55) as `{ cfg.pin_rd, cfg.pin_wr, cfg.pin_rs }`:

```cpp
    // either do this
    cfg.pin_ctrl = { TFT_RD, TFT_WR, TFT_RS };
    // or do this
    cfg_pin_rd = TFT_RD;
    cfg_pin_wr = TFT_WR;
    cfg_pin_rs = TFT_RS;
```

> and lastly before i hardwire all this in does lovyan support pushing images from psram? 

yes it does it transparently

**Reply by anatoli-dp** [reply]
Thanks very much for the quick reply. It is most helpful. Now I can confidently get back to finishing the design. It was late last night when I sent that and didn't read through everything so didn't realize it's a panel config issue for the cs line not a bus config, somehow missed that. But since u did reply is everything being set to my own defined pins have any negative speed drawbacks? I saw somewhere that the speed when going through the mix is around 20mhz max yet in the setup shows it can be 40mhz. Does this require specific pins or is it just whatever I choose to roll with without penalties

**Reply by tobozo** [reply]
40MHz is probably maximum **SPI** speed for the ili9341, but when using a parallel bus your mileage may vary; if unsure you can try at a low freq first (e.g. 4MHz) then increment until you get tearing, artifacts or blank screen (usually around 16MHz)

**Reply by anatoli-dp** [reply]
ili9341 can handle 60mhz reliably and can be iffy on 80mhz depending on how its done. As lone as u dont send the entire frame u can do 80mhz with data "blocks" like if u tile the display and then send over each tile as a seperate transaction
